discord/ext/commands/parameters.py

```diff
@@ -0,0 +1,238 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+from __future__ import annotations
+
+import inspect
+from operator import attrgetter
+from typing import TYPE_CHECKING, Any, Literal, Optional, OrderedDict, Union
+
+from discord.utils import MISSING, maybe_coroutine
+from . import converter
+from .errors import MissingRequiredArgument
+
+if TYPE_CHECKING:
+    from typing_extensions import Self
+
+    from discord import (
+        TextChannel,
+        Guild,
+        Member,
+        User,
+    )
+    from .context import Context
+
+__all__ = (
+    "Parameter",
+    "parameter",
+    "param",
+    "Author",
+    "CurrentChannel",
+    "CurrentGuild",
+)
+
+
+ParamKinds = Union[
+    Literal[inspect.Parameter.POSITIONAL_ONLY],
+    Literal[inspect.Parameter.POSITIONAL_OR_KEYWORD],
+    Literal[inspect.Parameter.VAR_POSITIONAL],
+    Literal[inspect.Parameter.KEYWORD_ONLY],
+    Literal[inspect.Parameter.VAR_KEYWORD],
+]
+
+empty: Any = inspect.Parameter.empty
+
+
+def _gen_property(name: str) -> property:
+    return property(
+        attrgetter(f"_{name}"),
+        lambda self, value: setattr(self, "_{name}", value),
+        doc="The parameter's {name}.",
+    )
+
+
+class Parameter(inspect.Parameter):
+    def __init__(
+        self,
+        name: str,
+        kind: ParamKinds,
+        default: Any = empty,
+        annotation: Any = empty,
+        displayed_default: str = empty,
+    ) -> None:
+        super().__init__(name=name, kind=kind, default=default, annotation=annotation)
+        self._name = name
+        self._kind = kind
+        self._default = default
+        self._annotation = annotation
+        self._displayed_default = displayed_default
+
+    def replace(
+        self,
+        *,
+        name: str = MISSING,  # MISSING here cause empty is valid
+        kind: ParamKinds = MISSING,
+        default: Any = MISSING,
+        annotation: Any = MISSING,
+        displayed_default: Any = MISSING,
+    ) -> Self:
+        if name is MISSING:
+            name = self._name
+        if kind is MISSING:
+            kind = self._kind  # type: ignore  # this assignment is actually safe
+        if default is MISSING:
+            default = self._default
+        if annotation is MISSING:
+            annotation = self._annotation
+        if displayed_default is MISSING:
+            displayed_default = self._displayed_default
+
+        return self.__class__(
+            name=name,
+            kind=kind,
+            default=default,
+            annotation=annotation,
+            displayed_default=displayed_default,
+        )
+
+    if not TYPE_CHECKING:  # this is to prevent anything breaking if inspect internals change
```

Comment:
I think as a consequence this is storing double the amount of attributes? The internal `inspect.Parameter` attributes and then the recreations in `Parameter`? So this would make Command objects take more memory. I'm sure this doesn't matter since most people aren't storing that many of these objects but it is unfortunate.

Maybe `inspect.Parameter` is slotted and this could use the same treatment? Unsure.

Reply:
This only storing the instance attributes from Parameter once is it not (Parameter has them called _name, _default etc)? Adding slots seems like a good idea